File: common/Seremis/SoulCraft/api/util/Coordinate3D.java
package Seremis.SoulCraft.api.util;

import net.minecraft.entity.Entity;
import net.minecraft.tileentity.TileEntity;

public class Coordinate3D implements Cloneable {

    public double x = 0, y = 0, z = 0;

    public Coordinate3D() {

    }

    public Coordinate3D(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Coordinate3D(Entity ent) {
        this.x = ent.posX;
        this.y = ent.posY;
        this.z = ent.posZ;
    }

    public Coordinate3D(TileEntity tile) {
        this.x = tile.xCoord;
        this.y = tile.yCoord;
        this.z = tile.zCoord;
    }

    public double getXCoord() {
        return this.x;
    }

    public double getYCoord() {
        return this.y;
    }

    public double getZCoord() {
        return this.z;
    }

    public void setCoords(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public void setCoords(float x, float y, float z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }
    
    public void setCoords(TileEntity tile) {
        this.x = tile.xCoord;
        this.y = tile.yCoord;
        this.z = tile.zCoord;
    }

    public void move(double x, double y, double z) {
        this.x =+x;
        this.y =+y;
        this.z =+z;
    }
    
    public int[] toArray() {
        int[] array = {(int) x, (int) y, (int) z};
        return array;
    }
    
    public Coordinate3D fromArray(double[] coords) {
        x = coords[0];
        y = coords[1];
        z = coords[2];
        return this;
    }
    
    public Coordinate3D fromArray(int[] coords) {
        x = coords[0];
        y = coords[1];
        z = coords[2];
        return this;
    }
    
    public double getDistanceTo(Coordinate3D coordinate) {
        Line3D line = new Line3D(this, coordinate);
        return line.getLength();
    }

    public Coordinate3D clone() {
        return new Coordinate3D(x, y, z);
    }
    
    @Override
    public boolean equals(Object o) {
        if(o instanceof Coordinate3D) {
            return equals((Coordinate3D)o);
        }
        return false;
    }
    
    public boolean equals(Coordinate3D coord) {
        if(x == coord.x && y == coord.y && z == coord.z)
            return true;
        return false;
    }
    
    public boolean equals(double x, double y, double z) {
        return equals(new Coordinate3D(x, y, z));
    }
    
    @Override
    public String toString() {
        return "Coordinate3D[x: "+x+" y: "+y+" z: "+z+"]";
    }
}

File: src/main/scala/seremis/geninfusion/api/util/vector/Coordinate3D.scala
package seremis.geninfusion.api.util.vector

import net.minecraft.entity.Entity
import net.minecraft.tileentity.TileEntity

class Coordinate3D(var x: Double, var y: Double, var z: Double) {

    def this(entity: Entity) {
        this(entity.posX, entity.posY, entity.posZ)
    }

    def this(tile: TileEntity) {
        this(tile.xCoord, tile.yCoord, tile.zCoord)
    }

    def getXCoord: Double = x
    def getYCoord: Double = y
    def getZCoord: Double = z

    def setXCoord(x: Double) = this.x = x
    def setYCoord(y: Double) = this.y = y
    def setZCoord(z: Double) = this.z = z

    def setCoord(x: Double, y: Double, z: Double) {
        this.x = x
        this.y = y
        this.z = z
    }

    def move(x: Double, y: Double, z: Double): Coordinate3D = {
        this.x += x
        this.y += y
        this.z += z
        this
    }

    def move(coord: Coordinate3D): Coordinate3D = move(coord.x, coord.y, coord.z)

    def moveBack(x: Double, y: Double, z: Double): Coordinate3D = {
        this.x -= x
        this.y -= y
        this.z -= z
        this
    }

    def moveBack(coord: Coordinate3D): Coordinate3D = moveBack(coord.x, coord.y, coord.z)

    def toArray: Array[Double] = Array(x, y, z)

    def fromArray(coords: Array[Double]): Coordinate3D = {
        x = coords(0)
        y = coords(1)
        z = coords(2)
        this
    }

    def getDistanceTo(coordinate: Coordinate3D): Double = {
        val line = new Line3D(this, coordinate)
        line.getLength
    }

    override def clone(): Coordinate3D = new Coordinate3D(x, y, z)

    override def equals(o: Any): Boolean = {
        if (o.isInstanceOf[Coordinate3D]) {
            val coord = o.asInstanceOf[Coordinate3D]
            return coord.x == x && coord.y == y && coord.z == z
        }
        if (o.isInstanceOf[TileEntity]) {
            return equals(new Coordinate3D(o.asInstanceOf[TileEntity]))
        }
        if (o.isInstanceOf[Entity]) {
            return equals(new Coordinate3D(o.asInstanceOf[Entity]))
        }
        false
    }

    override def toString: String = "Coordinate3D[x: " + x + " y: " + y + " z: " + z + "]"
}
